File: towel_detection/pcl_ros_processing/src/send_data.cpp
#include <ros/ros.h>
#include "std_msgs/Bool.h"
#include <geometry_msgs/Vector3.h>
#include "serial/serial.h"

using std::exception;
using std::cout;
using std::cerr;
using std::endl;

//通信設定
const std::string port = "/dev/serialUSB";
const uint32_t baudrate=115200;
serial::Timeout timeout(serial::Timeout::max(), 100,0,0,0);//read:100ms, write:0ms
serial::Serial ser;


// serial::Serial ser;//(port, baudrate,timeout);

//変数定義
geometry_msgs::Vector3 now_destination;

ros::Time last_time;
bool has_flushed_recieve_buffer = false;
bool now=false;

ros::Publisher pub_send;

//コールバック関数--------------------------------------------
void cb_destination(const geometry_msgs::Vector3& destination)
{
  now_destination = destination;
  now = true;
}

//--------------------------------------------------------
int main(int argc, char *argv[])
{
  ros::init(argc, argv, "send_data");
  ros::NodeHandle nh;
  ros::NodeHandle pn("~");

  pub_send = nh.advertise<std_msgs::Bool> ("no_send",10);
  ros::Publisher pub_omlette = nh.advertise<std_msgs::Bool> ("omelette",10);

  ros::Subscriber sub_destination = nh.subscribe("destination",10,cb_destination);
  // ros::Subscriber sub_request = nh.subscribe("send",10,cb_request);

  std::string send_data = "n\n";
  std_msgs::Bool removed, omelette_flag;

  //param
  int loop_rate,plus,request_timeout;
  pn.param("loop_rate",loop_rate,50);
  pn.param("plus",plus,160);
  pn.param("request_timeout",request_timeout,10);

  ser.setPort(port);
  ser.setTimeout(timeout);
  ser.setBaudrate(baudrate);

  //接続待ち
  while (not ser.isOpen()) {
    try{
      ser.open();
    } catch (exception &e) {
      ROS_WARN("can't conect");
      ros::Duration(0.5).sleep();
    }
  }

  ros::Rate loop_rate_(loop_rate);
  while (ros::ok()) {
    ros::Time now_time = ros::Time::now();
    if (ser.getCTS()) {
      if(not has_flushed_recieve_buffer){
        ser.flushInput();
        has_flushed_recieve_buffer = true;
      }

      if (now)
      {
        if (now_destination.z == 0)
        {
          send_data = "x" + std::to_string(int(now_destination.y * 1000)) + "\n";
          ROS_INFO("move destination x:%f", now_destination.y * 1000);
          removed.data = false;
        }

        else if (now_destination.z == 1)
        {
          send_data = "p" + std::to_string(int(now_destination.x * 1000) + plus) + "\n";
          ROS_INFO("Pull down destination y:%f", now_destination.x * 1000);
          removed.data = false;
        }

        else if (now_destination.z == 2)
        {
          send_data = "c" + std::to_string(int(now_destination.x * 1000) + plus) + "\n";
          ROS_INFO("Catch destination y:%f", now_destination.x * 1000);
          removed.data = false;
        }

        else if (now_destination.z == 3)
        {
          send_data = "n\n";
          ROS_INFO("Ignore!");
          removed.data = false;
        }

        else if (now_destination.z == 4)
        {
          send_data = "r" + std::to_string(int(now_destination.x * 1000) + plus + 30) + "\n";
          ROS_INFO("Remove!");
          removed.data = true;
        }

        else
        {
          ROS_INFO("send error (destination.z:%f is not available", now_destination.z);
          send_data = "n\n";
          removed.data = false;
        }
        now = false;
        ser.write(send_data);
        ROS_INFO_STREAM(send_data);
        //pub_send.publish(removed);
        last_time = ros::Time::now();
      }
      
    }

    //cts==false
    else if ((now_time-last_time).toSec() > request_timeout) {
      has_flushed_recieve_buffer = false;
      if (ser.available())
      {
        std::string mode = ser.readline(); //65536,"\n"

        if (mode == "o\n")
        {
          ROS_INFO("Execute Omelette");
          omelette_flag.data = true;
        }

        else if (mode == "n\n")
        {
          ROS_INFO("Not Execute Omelette");
          omelette_flag.data = false;
        }

        else
        {
          ROS_ERROR("mode error");
          omelette_flag.data = false;
        }
        pub_omlette.publish(omelette_flag);
      }
    }

    loop_rate_.sleep();
    ros::spinOnce();
  }
}

File: towel_detection/pcl_ros_processing/scripts/send_data.py
#!/usr/bin/env python

import rospy
import serial
import math
from std_msgs.msg import String
from std_msgs.msg import Bool
from geometry_msgs.msg import Vector3

ser = serial.Serial(port="/dev/ttyUSB0", baudrate=115200,
                    rtscts=True, timeout=0.1)

send_data = "n" + "\n"
last_catch = False
has_flushed_recieve_buffer = False
omelette_flag = Bool()
omelette_flag.data = False


def cb_destination(destination):
    global send_data
    global last_catch
    global omelette_flag
    global has_flushed_recieve_buffer
    plus = 160

    catched = Bool()

    if ser.cts:
        if not has_flushed_recieve_buffer:
            ser.reset_input_buffer()
            has_flushed_recieve_buffer = True

        if destination.z == 0:
            send_data = "x" + str(int(math.floor(destination.y*1000))) + "\n"
            rospy.loginfo("move destination x:%f", destination.y)
            catched.data = False

            if abs(destination.y) >= 0.12:
                last_catch = False

        elif destination.z == 1:
            if last_catch:
                send_data = "c" + \
                    str(int(math.floor(destination.x*1000))+plus) + "\n"
                rospy.loginfo("Catch destination y:%f", destination.x)
                last_catch = False

            else:
                send_data = "p" + \
                    str(int(math.floor(destination.x*1000))+plus) + "\n"
                rospy.loginfo("Pull down destination y:%f", destination.x)
                last_catch = False

        elif destination.z == 2:
            send_data = "c" + \
                str(int(math.floor(destination.x*1000))+plus) + "\n"
            rospy.loginfo("Catch destination y:%f", destination.x)
            last_catch = False

        elif destination.z == 3:
            send_data = "n" + "\n"
            rospy.loginfo("Ignore!")
            last_catch = False

        elif destination.z == 4:
            send_data = "r" + \
                str(int(math.floor(destination.x*1000))+plus+30) + "\n"
            rospy.loginfo("Remove!")
            last_catch = False
            #catched.data = True

        else:
            rospy.loginfo(
                "send error (destination.z:%f is not available", destination.z)
            send_data = "n" + "\n"

        ser.write(send_data.encode())
        print(send_data)


def cb_send(send):
    global omelette_flag
    global has_flushed_recieve_buffer

    pub = rospy.Publisher('omelette', Bool, queue_size=10)

    if not ser.cts:
        has_flushed_recieve_buffer = False
        if ser.in_waiting:
            mode = ser.readline()

            if mode.decode() == "o\n":
                print("Execute Omelette")
                omelette_flag.data = True

            elif mode.decode() == "n\n":
                print("Not Execute Omelette")
                omelette_flag.data = False

            else:
                rospy.logerr("mode error")
                omelette_flag.data = False

            pub.publish(omelette_flag)


def listner():
    rospy.init_node('listner', anonymous=True)

    rospy.Subscriber('destination', Vector3, cb_destination)
    rospy.Subscriber('send', Bool, cb_send)

    rospy.loginfo("start send_data node")

    rospy.spin()


if __name__ == '__main__':
    listner()
